### PR TITLE
Add restricted controller messenger

### PR DESCRIPTION
**Pull request by Gudahtt**
A method has been added to the controller messenger that returns a _restricted_ instance of the controller messenger. This instance is restricted to registering actions and events within a particular namespace, typically being the name of the controller. Access to event subscriptions and calling actions is also restricted based on the given allowlist for each.

This is meant to address feedback left on the [Controller Messaging System proposal](https://www.notion.so/Controller-Messaging-System-617efb02b9e54bd0a0b0e44c6f776d85).

### CODE REVIEW

**Comment by Gudahtt** [comment]
This depends upon #377

**Comment by Gudahtt** [comment]
My objective with this restricted messaging system was to prevent _accidental_ use of methods beyond the intended access. I'm not protecting against access to private methods/properties, or anything sneaky like that. We have lint rules and types to protect against us doing that. This approach isn't adequate for dealing with untrusted code. I don't think we'll be using the controller messaging system for untrusted code though - at least not directly.

The snaps API will likely use the messaging system, but it will use other means to prevent unintended access (e.g. an RPC layer, SES, the permission system, etc.).

File: src/ControllerMessenger.ts [hunk 1]
```diff
@@ -158,4 +247,17 @@ export class ControllerMessenger<Action extends ActionConstraint, Event extends
   clearSubscriptions() {
     this.events.clear();
   }
+
+  getRestricted<N extends string, AllowedAction extends string, AllowedEvent extends string>(
```
Comment (Gudahtt) [inline src/ControllerMessenger.ts:251]: All of these new methods need docs.

Also I'm going to change these params to take an options object instead, to improve readability.

Comment (Gudahtt) [reply src/ControllerMessenger.ts]: Done.

File: src/BaseControllerV2.test.ts [hunk 2]
```diff
@@ -504,4 +564,139 @@ describe('getPersistentState', () => {
...
+          return { maxVisitors: max };
+        });
+      };
+
+      destroy() {
+        super.destroy();
+      }
+    }
+
+    it('should allow messaging between controllers', () => {
```
Comment (Gudahtt) [inline src/BaseControllerV2.test.ts:675]: This is the most complete example so far. It demonstrates calling an action and subscribing to an event in another controller.

File: src/BaseControllerV2.test.ts [hunk 3]
```diff
@@ -511,4 +571,149 @@ describe('getPersistentState', () => {
...
+      const controllerMessenger = new ControllerMessenger<
+        VisitorControllerAction | VisitorOverflowControllerAction,
+        VisitorControllerEvent | VisitorOverflowControllerEvent
+      >();
+      const visitorControllerMessenger = controllerMessenger.getRestricted({
+        name: 'VisitorController',
+        allowedActions: [],
+        allowedEvents: [],
+      });
+      const visitorController = new VisitorController(visitorControllerMessenger);
```
Comment (danfinlay) [inline src/BaseControllerV2.test.ts:702]: I thought we'd moved to an options bag for this type of pattern, is that optional?

File: src/BaseControllerV2.test.ts [hunk 4]
```diff
@@ -511,4 +571,149 @@ describe('getPersistentState', () => {
...
+      const visitorOverflowControllerMessenger = controllerMessenger.getRestricted({
+        name: 'VisitorOverflowController',
+        allowedActions: ['VisitorController:clear'],
+        allowedEvents: ['VisitorController:stateChange'],
+      });
+      const visitorOverflowController = new VisitorOverflowController(visitorOverflowControllerMessenger);
+
+      controllerMessenger.call('VisitorOverflowController:updateMax', 2);
```
Comment (danfinlay) [inline src/BaseControllerV2.test.ts:710]: Not a strong opinion, but I slightly feel like the method params should be an array as a second argument. I don't feel this strongly enough to request you do it, let alone block, but feel it just enough to want to record that twinge here in case anyone else feels more strongly.

**Review by danfinlay — APPROVED** [review]
Overall, I think this is good to go. I left two comment/observations, but neither needs to block, I'm happy with how this works and am excited to have it available.

File: src/BaseControllerV2.test.ts (hunk 3 above)
Comment (Gudahtt) [reply src/BaseControllerV2.test.ts:702]: The base controller and the messaging system both use the options bag pattern. There are no constraints upon controller constructors - I don't think it's possible for us to constrain that :thinking:.

This was just a mock controller, which I kept as simple as possible to reduce the size of this implementation (which ended up being rather large anyway, as you can see). I was also thinking that if we did have a controller with zero options, it would be OK for it to accept the messaging system directly without an options bag. I could update it to use an options bag parameter anyway though, for consistency.

File: src/BaseControllerV2.test.ts (hunk 4 above)
Comment (Gudahtt) [reply src/BaseControllerV2.test.ts:710]: This is interesting :thinking: It would have been slightly easier for me to get the types working if this accepted an array. I had put the extra effort into unwrapping them because I thought nesting arguments in an array would make this more difficult to use. This felt to me more like a "normal" method.

It also allows us to do this:

```suggestion
      const updateMax = controllerMessenger.bind(controllerMessenger, 'VisitorOverflowController:updateMax');
      updateMax(2);
```
So that we can create an instance of the action that is _exactly_ like the function at the other end, except that it's facilitated by the messenger system.

File: src/ControllerMessenger.test.ts [hunk 5]
```diff
@@ -289,3 +289,396 @@ describe('ControllerMessenger', () => {
...
+    const restrictedControllerMessenger = controllerMessenger.getRestricted({
+      name: 'MessageController',
+      allowedActions: [],
+      allowedEvents: ['MessageController:message'],
+    });
+
+    expect(() => restrictedControllerMessenger.clearEventSubscriptions('MessageController:message')).not.toThrow();
+  });
+
+  it('should allow calling an external action', () => {
```
Comment (rekmarks) [inline src/ControllerMessenger.test.ts:640]: nit: It _might_ be worth adding test cases that mix external and internal (relative to the namespace) actions and events.

For instance, we could copy the below case and modify it to have `allowedActions: ['CountController:count', 'OtherController:foo']`. Not necessary with respect to the implementation, just to cover an input edge case.

File: src/BaseControllerV2.test.ts [hunk 6]
```diff
@@ -511,4 +571,149 @@ describe('getPersistentState', () => {
 
     expect(persistentState).toEqual({ count: 1 });
   });
+
+  describe('inter-controller communication', () => {
```
Comment (rekmarks) [inline src/BaseControllerV2.test.ts:575]: nit: It'd be nice if we had some breathing space between the type declarations in here.

File: src/BaseControllerV2.test.ts (hunk 2 above)
Comment (rekmarks) [reply src/BaseControllerV2.test.ts:692]: I think I love it. 😍

**Review by rekmarks — CHANGES_REQUESTED** [review]
A bunch of nits, but overall I think this rules.

File: src/ControllerMessenger.ts [hunk 7]
```diff
@@ -14,6 +14,205 @@ type ExtractEventPayload<Event, T> = Event extends { type: T; payload: infer P }
...
+   *
+   * The action type this handler is registered under *must* be in the current namespace.
+   *
+   * @param actionType - The action type. This is a unqiue identifier for this action.
+   * @param handler- The action handler. This function gets called when the `call` method is
+   *   invoked with the given action type.
+   * @throws Will throw when a handler has been registered for this action type already.
+   */
+  registerActionHandler<T extends Namespaced<N, Action['type']>>(action: T, handler: ActionHandler<Action, T>) {
+    /* istanbul ignore if */
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/ControllerMessenger.ts:93]: I added ignore statements for any lines that were impossible to test in TypeScript. The types here statically prevent ever passing in an action string that doesn't start with the controller name - it wouldn't compile.

[the remark this answers is not in the document]
Comment (rekmarks) [reply src/ControllerMessenger.ts:93]: Gah! Naturally. In the past I've added `.js` tests to try input mangling, but this should be fine. Can you put the justification in the `istanbul` comment?

[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/ControllerMessenger.ts]: Done in 65c672149fb7489b3595cd9e1cbe9f5747bdc073

File: src/ControllerMessenger.test.ts (hunk 5 above)
Comment (Gudahtt) [reply src/ControllerMessenger.test.ts:640]: Done in deb2409d48aa7d966eaa241a408d9afd5522731d

File: src/BaseControllerV2.test.ts (hunk 6 above)
Comment (Gudahtt) [reply src/BaseControllerV2.test.ts:575]: Done in d25bb8e692d21cdb7ccd00492c1e510ec5942c5e

I left the type declarations together, because it still felt right to group them together. But now there are newlines around them, and around the metadata object.

File: src/ControllerMessenger.ts [hunk 8]
```diff
@@ -14,6 +14,205 @@ type ExtractEventPayload<Event, T> = Event extends { type: T; payload: infer P }
...
+export type Namespaced<Name extends string, T> = T extends `${Name}:${string}` ? T : never;
+
+/**
+ * A restricted controller messenger.
+ *
+ * This acts as a wrapper around the controller messenger instance that restricts access to actions
+ * and events.
+ */
+export class RestrictedControllerMessenger<
+  N extends string,
```
[the remark this answers is not in the document]
Comment (Gudahtt) [reply src/ControllerMessenger.ts:31]: I'll address this in a follow-up PR

**Comment by Gudahtt** [comment]
This has been rebased, and all comments have been addressed.

**Review by rekmarks — APPROVED** [review]
Let's gooo!
